### PR TITLE
Issue 4795 - include ref and contextdir in github links

### PR DESCRIPTION
**Pull request by jwforres**

### CODE REVIEW

**Comment by jwforres** [comment]
Fixes https://github.com/openshift/origin/issues/4795

@liggitt

File: assets/app/scripts/filters/util.js [hunk 1]
```diff
@@ -156,12 +156,27 @@ angular.module('openshiftConsole')
...
+            contextDir = contextDir.substring(1);
+          }
+          // make sure its not an empty string now that we trimmed /
+          if (contextDir) {
+            // do this here so we aren't adding /tree/master if we don't have a contextDir
+            if (!ref || ref === "master") {
+              link += "/tree/master";
+            }
+            link += "/" + contextDir;
+          }
```
Comment (spadgett) [inline assets/app/scripts/filters/util.js]: I think this is simpler, but does the same thing (untested):

```
        if (contextDir) {
            contextDir = contextDir.substring(1);
        }

        if (contextDir) {
          link += "/tree/" + (ref || "master") + "/" + contextDir;
        } else if (ref || ref !== "master" {
          link += "/tree/" + ref;
        }
```

File: assets/app/scripts/filters/util.js [hunk 2]
```diff
@@ -156,12 +156,27 @@ angular.module('openshiftConsole')
     };
   })
   .filter('githubLink', function() {
-    return function(link, commit) {
+    return function(link, ref, contextDir) {
       var m = link.match(/^(?:https?:\/\/|git:\/\/|git\+ssh:\/\/|git\+https:\/\/)?(?:[^@]+@)?github\.com[:\/]([^\/]+\/[^\/]+?)(?:\.git(#.*)?)?$/);
       if (m) {
         link = "https://github.com/" + m[1];
-        if (commit) {
-          link += "/commit/" + commit;
```
Comment (spadgett) [inline assets/app/scripts/filters/util.js]: Do we need to check if link ends in `/` ? (I see that test wasn't there before.)

**Comment by spadgett** [comment]
Call `encodeURIComponent` on all elements we add to the URL.

File: assets/app/scripts/filters/util.js (hunk 2 above)
Comment (jwforres) [inline assets/app/scripts/filters/util.js]: The existing regex is wrong, it doesnt match if the URL has a trailing slash, i'll fix that and make sure that m[1] doesn't include the /

**Comment by jwforres** [comment]
encodeURIComponent will encode the / chars in the contextDir path, so im calling it but then replacing %2F with / to maintain the paths in the URL

**Comment by jwforres** [comment]
@spadgett updated

**Comment by spadgett** [comment]
LGTM

**Comment by jwforres** [comment]
[merge]

**Comment by jwforres** [comment]
[merge]

**Pull request merged**